eis/run_models.py

```diff
@@ -413,6 +413,75 @@ def evaluations(self, predictions_proba, predictions_binary, test_y, model_id, t
         db_conn.close()
         return None
 
+    def individual_feature_ranking(self, fitted_model, test_matrix, model_id, test_date):
+        ###################
+        # This method is a beta version tested for top k optimized random forests
+        ###################
+
+        # get the list of all features in the test matrix
+        feature_list = test_matrix.columns
+
+        # extract all global feature importance values from the tree
+        importance_dict = {}
+        for j, value in enumerate(fitted_model.feature_importances_):
+            if value > 0:
+                importance_dict[feature_list[j]] = value
+
+        # filter out dummies
+        importance_dict_filtered = {k: v for k, v in importance_dict.items() if not 'dummy' in k}
+
+        rftree_feature_list = sorted(importance_dict_filtered, key=importance_dict_filtered.get, reverse=True)
+
+        # create a new text matrix with float conversion (panda stores some columns from postgres as decimal object)
+        tmp_test = test_matrix.applymap(lambda x: float(x))
+        test_matrix_reduced = pd.DataFrame()
+
+        # add the top 30 features from the RandomForest used without dummies
+        for idx, feature in enumerate(rftree_feature_list):
+            if idx >= 30:
+                break
+            test_matrix_reduced[feature] = tmp_test[feature]
+
+        test_matrix_rank_distance = pd.DataFrame()
+
+        for feature in test_matrix_reduced.columns:
+            feature_median = test_matrix_reduced[feature].median()
+            test_matrix_reduced[feature + '_rank'] = test_matrix_reduced[feature].rank()
+            # the index of the median value
+            idx = np.nanargmin(np.abs(test_matrix_reduced[feature] - feature_median))
+            median_rank = test_matrix_reduced[feature].iloc[idx]
+            test_matrix_rank_distance[feature] = np.abs(test_matrix_reduced[feature + '_rank'] - median_rank)
+
+        test_matrix_rank_distance = test_matrix_rank_distance.reset_index(drop=True)
+
+        # find the top 5 features where the distance to the median rank value (for standardisation) is the highest
+        # across the top 30 features
+        matrix_transposed = test_matrix_rank_distance.T
+        result = pd.DataFrame(np.zeros((0, 5)), columns=['risk_1', 'risk_2', 'risk_3', 'risk_4', 'risk_5'])
+        for i in matrix_transposed.columns:
```

Comment:
You don't need to concatenate (is not the recommended way of doing this), you could create a list of tuples and then creating the pandas `DataFrame` (and you could remove the previous comment... but this needs more careful thinking, so you could leave this code here in this version.

For reference see: https://stackoverflow.com/questions/28056171/how-to-build-and-fill-pandas-dataframe-from-for-loop